src/Compilers/Core/Portable/SourceGeneration/Nodes/PredicateSyntaxStrategy.cs

```diff
@@ -84,14 +84,24 @@ public void SaveStateAndFree(StateTableStore.Builder tables)
                     {
                         var stopwatch = SharedStopwatch.StartNew();
                         var nodes = getFilteredNodes(root.Value, _owner._filterFunc, cancellationToken);
-                        entry = _filterTable.AddEntries(nodes, state, stopwatch.Elapsed, noInputStepsStepInfo, state);
+
+                        if (state != EntryState.Modified || !_filterTable.TryModifyEntries(nodes, Roslyn.Utilities.ReferenceEqualityComparer.Instance, stopwatch.Elapsed, noInputStepsStepInfo, state, out entry))
+                        {
+                            entry = _filterTable.AddEntries(nodes, state, stopwatch.Elapsed, noInputStepsStepInfo, state);
+                        }
                     }
 
                     // now, using the obtained syntax nodes, run the transform
-                    foreach (SyntaxNode node in entry)
+                    for (var i = 0; i < entry.Count; i++)
```

Comment:
This portion of the change matches the description, but I didn't quite follow the portion above (with `TryModifyEntries`). Let's try to find some time to sync so that I can understand

Reply:
This is why I think it's correct: this syntax node implementation is basically `Where` + `Select`, only implemented in terms of the low-level `NodeStateTable<T>.Builder`. But if you look at how `Where` and `Select` are implemented they need to call `TryUseCachedEntries`, `TryModifyEntries`, `AddEntries`, `TryRemoveEntries` in similar way on the builder to handle all cases. Only the syntax node implementation wasn't doing that. (`Where` and `Select` both use `TransformNode`, but other nodes also call the same methods on the builder in the same way.)

https://github.com/dotnet/roslyn/blob/a54f0b411285fc5949f64da2c1bec9aeeebe46f7/src/Compilers/Core/Portable/SourceGeneration/Nodes/TransformNode.cs#L67-L81

Reply:
From offline chat, please confirm that the test did indeed cover both changes